src/array.vala

```diff
@@ -464,4 +396,175 @@ public class Vast.Array : Object
+         *
+         * the final build array is built after this.end() is called.
+         *
+         * methods can be chained or called in a loop.
+         *
+         */
+        private size_t shape[32];
+        private ssize_t strides[32];
+        private size_t origin;
+        private Array array;
+        private size_t dimension;
+
+        internal Builder(Array array, size_t dimension)
+        {
+            assert (dimension >= array._dimension);
+            for (var i = 0; i < array._dimension; i ++) {
+                shape[i] = array._shape[i];
+                strides[i] = array._strides[i];
+            }
+            for (var i = array._dimension; i < dimension; i ++) {
+                shape[i] = 1;
+                strides[i] = 0;
+            }
+            origin = array.origin;
+            this.array = array;
+            this.dimension = dimension;
+        }
+
+        /* from:to:step */
+        public Builder
+        slice(ssize_t axis, ssize_t from, ssize_t to, ssize_t step=1)
+        {
+            return qslice(axis, (ssize_t[]) &from, (ssize_t[]) &to, step);
+        }
+
+        /* :to:step */
+        public Builder
+        head(ssize_t axis, ssize_t to, ssize_t step=1)
+        {
+            return qslice(axis, null, (ssize_t[]) &to, step);
+        }
+
+        /* from::step */
+        public Builder
+        tail(ssize_t axis, ssize_t from, ssize_t step=1)
+        {
+            return qslice(axis, (ssize_t[]) &from, null, step);
+        }
+
+        /* ::step */
+        public Builder
+        step(ssize_t axis, ssize_t step=1)
+        {
+            return qslice(axis, null, null, step);
+        }
+
+        public Builder
+        qslice(ssize_t axis,
+            [CCode (array_length = false)]
+            ssize_t []? qfrom = null,
+            [CCode (array_length = false)]
+            ssize_t []? qto = null,
+            ssize_t step = 1)
+        {
+            axis = wrap_by_dimension(axis);
+
+            ssize_t from, to;
+            if(qfrom == null)
+                if (step > 0)
+                    from = 0;
+                else
+                    from = (ssize_t) shape[axis] - 1;
+            else {
+                from = qfrom[0];
+                from = from < 0 ? (ssize_t) shape[axis] + from : from;
+            }
+
+            if(qto == null)
+                if (step > 0)
+                    to = (ssize_t) shape[axis];
+                else
+                    to = -1;
+            else {
+                to = qto[0];
+                to = to < 0 ? (ssize_t) shape[axis] + to   : to;
+            }
+
+            origin += from * strides[axis];
+
+            /* XXX: hopefully this rounds down even for negative steps */
+            shape[axis] = (to - from) / step;
+
+            strides[axis] *= step;
+
+            assert (shape[axis] >= 0);
+            return this;
+        }
+
+        public Builder
+        reset_dimension(size_t dimension)
```

Comment:
Is this really useful? Maybe have a `revert` method to return the original array instead?

```vala
arr.build ()
    .slice (0, ...)
    .revert ();
```